compiler/passes/src/const_propagation/const_propagator.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright (C) 2019-2024 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{SymbolTable, TypeTable};
+
+use leo_ast::{Expression, IntegerType, Literal, NodeBuilder, NodeID, TupleExpression};
+use leo_errors::{StaticAnalyzerError, emitter::Handler};
+use leo_interpreter::Value;
+use leo_span::{Span, Symbol};
+
+/// A pass to perform const propagation and folding.
```

Comment:
It would be helpful to document the cases where we intentionally do not propagate constants, so that users know the patterns the compiler is looking for.